test/middleware.js

```diff
@@ -314,4 +315,91 @@ describe('middleware', function() {
       });
     });
   });
+
+  describe('extra information (x)', function() {
+    var connection;
+    var db;
+    var doc;
+
+    beforeEach(function(done) {
+      connection = this.backend.connect();
+      db = this.backend.db;
+      doc = connection.get('dogs', 'fido');
+
+      doc.create({name: 'fido'}, done);
+      // Need to actively enable this feature
+      doc.submitSource = true;
+    });
+
+    it('has the source in commit middleware', function(done) {
+      this.backend.use('commit', function(request) {
+        expect(request.extra).to.eql({source: 'trainer'});
+        done();
+      });
+
+      doc.submitOp([{p: ['tricks'], oi: ['fetch']}], {source: 'trainer'}, errorHandler(done));
+    });
+
+    it('has the source in afterWrite middleware', function(done) {
+      this.backend.use('afterWrite', function(request) {
+        expect(request.extra).to.eql({source: 'trainer'});
+        done();
+      });
+
+      doc.submitOp([{p: ['tricks'], oi: ['fetch']}], {source: 'trainer'}, errorHandler(done));
+    });
+
+    it('does not commit extra information to the database', function(done) {
+      doc.submitOp([{p: ['tricks'], oi: ['fetch']}], {source: 'trainer'}, function(error) {
+        if (error) return done(error);
+        var ops = db.ops.dogs.fido;
+        ops.forEach(function(op) {
+          expect('x' in op).to.be.false;
+        });
+        done();
+      });
+    });
+
+    it('does not submit the source if it is disabled', function(done) {
+      doc.submitSource = false;
+
+      this.backend.use('commit', function(request) {
+        expect('source' in request.extra).to.be.false;
+        done();
+      });
+
+      doc.submitOp([{p: ['tricks'], oi: ['fetch']}], {source: 'trainer'}, errorHandler(done));
+    });
+
+    it('composes ops with the same source', function(done) {
```

Comment:
Might be good to add a test for 'composes ops with different sources if `submitSource` is false' though not sure how likely that scenario is

Reply:
Done.